### PR TITLE
Refactor navigation components for easier access restriction

### PR DESCRIPTION
**Pull request by marcellp**
# Background
#### Link to issue 
https://biomage.atlassian.net/browse/BIOMAGE-483

#### Link to staging deployment URL 
https://ui-experimentid-global.scp-staging.biomage.net/

#### Links to any Pull Requests related to this
N/A

#### Anything else the reviewers should know about the changes here
N/A

# Changes
### Code changes
Create a higher-level preload component that guards against loading a page before the experiment.

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [x] Unit tests written
- [x] Tested locally with Inframock
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [x] Relevant Github READMEs updated
- [x] Relevant wiki pages created/updated

### Approvers
- [x] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

**Comment by marcellp** [comment]
Staging now available at https://ui-experimentid-global.scp-staging.biomage.net/.

File: package.json [hunk 1]
```diff
@@ -64,7 +64,11 @@
     "terser-webpack-plugin": "^1.1.0",
     "thread-loader": "^2.1.3",
     "webpack": "^4.43.0",
-    "webpack-bundle-analyzer": "^3.3.2"
+    "webpack-bundle-analyzer": "^3.3.2",
+    "swr": "^0.3.2",
+    "prop-types": "^15.7.2",
+    "object-hash": "^2.0.3",
+    "localforage": "^1.7.4"
```
Comment (ivababukova) [inline package.json]: why were these moved?

File: src/pages/_app.jsx [hunk 2]
```diff
@@ -19,18 +21,67 @@ Router.events.on('routeChangeComplete', () => NProgress.done());
...
+
+  // Only hydrate pages when experiment ID is loaded.
+  useEffect(() => {
+    if (!router.route.includes('experimentId')) {
+      setExperimentId(false);
+    }
+
+    if (router.route.includes('experimentId')) {
+      setExperimentId(router.query.experimentId);
+      console.log(experimentId, 'changed');
```
Comment (ivababukova) [inline src/pages/_app.jsx]: remove the console

File: src/pages/_app.jsx [hunk 3]
```diff
@@ -19,18 +21,67 @@ Router.events.on('routeChangeComplete', () => NProgress.done());
...
       return <Component {...pageProps} />;
     }
 
-    if (!isBrowser) {
+    if (!experimentId && experimentId !== false) {
+      return (<PreloadContent />);
+    }
+
+    if (experimentId && !experimentData && !experimentError) {
+      console.log(experimentId, experimentData, experimentError);
```
Comment (ivababukova) [inline src/pages/_app.jsx]: remove this

File: src/pages/experiments/[experimentId]/data-exploration/index.jsx [hunk 4]
```diff
@@ -34,24 +33,29 @@ const renderWindow = (tile, width, height) => {
...
+const ExplorationViewPage = ({ experimentId, experimentData, route }) => {
   const dispatch = useDispatch();
-  const router = useRouter();
-  const { experimentId } = router.query;
   const layout = useSelector((state) => state.layout);
   const { windows, panel } = layout;
   const [selectedTab, setSelectedTab] = useState(panel);
+  const [addMenuVisible, setAddMenuVisible] = useState(false);
+
+  console.log(route);
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/data-exploration/index.jsx]: remove this

**Review by ivababukova — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (marcellp) [reply package.json]: Because `_app.jsx` needs to have all its dependencies inside `dependencies` and not `devDependencies`.

`swr`, `prop-types` are used now by `_app.jsx`. `object-hash` and `localforage` are now not in the list because I refactored the one piece of code where they were used and it's not really necessary anymore.

Comment (ivababukova) [reply package.json]: makes sense, thanks for explaining!
